From f10acc498d99f29bebebfd5cb20d42d3438b8d5d Mon Sep 17 00:00:00 2001
From: Xulin Sun <xulin.sun@windriver.com>
Date: Mon, 29 Jul 2019 17:06:36 +0800
Subject: [PATCH] usb: lan78xx: disable local interrupt for phyirq handler

To avoid below call trace:
[    5.763188] libphy: lan78xx-mdiobus: probed
[    6.759316] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
[    7.720332] -----------[ cut here ]-----------
[    7.723727] irq 151 handler irq_default_primary_handler+0x0/0x20 enabled interrupts
[    7.730199] WARNING: CPU: 3 PID: 95 at kernel/irq/handle.c:153 __handle_irq_event_percpu+0x288/0x2b8
[    7.730203] Modules linked in:
[    7.730220] CPU: 3 PID: 95 Comm: irq/9-dwc_otg_h Not tainted 4.18.40-rt8-yocto-preempt-rt #1
[    7.730224] Hardware name: Raspberry Pi 3 Model B+ (DT)
[    7.730232] pstate: 40000005 (nZcv daif -PAN -UAO)
[    7.730240] pc : __handle_irq_event_percpu+0x288/0x2b8
[    7.730248] lr : __handle_irq_event_percpu+0x288/0x2b8
[    7.730252] sp : ffffff80093ebab0
[    7.730256] x29: ffffff80093ebab0 x28: 0000000000000002
[    7.730267] x27: ffffff8008ed1018 x26: ffffff8008eed000
[    7.730278] x25: ffffffc0373bd800 x24: ffffff80093ebb54
[    7.730289] x23: ffffff8008fc8aa8 x22: ffffff8008fc8a80
[    7.730299] x21: 0000000000000097 x20: 0000000000000000
[    7.730310] x19: ffffffc036d90b80 x18: 0000000000000060
[    7.730320] x17: 0000000000000000 x16: ffffffc0372bbb80
[    7.730330] x15: ffffffffffffffff x14: 746e692064656c62
[    7.730340] x13: 616e652030327830 x12: 2f3078302b72656c
[    7.730351] x11: 646e61685f797261 x10: 6d6972705f746c75
[    7.730361] x9 : 616665645f717269 x8 : 2072656c646e6168
[    7.730372] x7 : 2031353120717269 x6 : 0000000000000003
[    7.730382] x5 : 0000000000000000 x4 : 0000000000000000
[    7.730391] x3 : 0000000000000000 x2 : ffffffffffffffff
[    7.730402] x1 : 93821033f0d7e500 x0 : 0000000000000000
[    7.730414] Call trace:
[    7.730422]  __handle_irq_event_percpu+0x288/0x2b8
[    7.730430]  handle_irq_event_percpu+0x68/0xb8
[    7.730437]  handle_irq_event+0x68/0xb8
[    7.730446]  handle_simple_irq+0xa8/0x108
[    7.730454]  generic_handle_irq+0x34/0x50
[    7.730463]  intr_complete+0xbc/0xf0
[    7.730475]  __usb_hcd_giveback_urb+0x5c/0xd0
[    7.730484]  usb_hcd_giveback_urb+0x10c/0x120
[    7.730493]  completion_tasklet_func+0x7c/0xd8
[    7.730502]  tasklet_callback+0x20/0x30
[    7.730513]  tasklet_action_common.isra.0+0x84/0x128
[    7.730522]  tasklet_hi_action+0x38/0x48
[    7.730530]  do_current_softirqs+0x1f4/0x3e8
[    7.730538]  __local_bh_enable+0x58/0x88
[    7.730547]  irq_forced_thread_fn+0x7c/0xa8
[    7.730554]  irq_thread+0x140/0x1a8
[    7.730562]  kthread+0x130/0x138
[    7.730571]  ret_from_fork+0x10/0x1c
[    7.730576] --[ end trace 0000000000000002 ]--

Local IRQs should be disabled to enter generic_handle_irq of PHY irq.

Signed-off-by: Xulin Sun <xulin.sun@windriver.com>
---
 drivers/net/usb/lan78xx.c | 6 +++++-
 1 file changed, 5 insertions(+), 1 deletion(-)

diff --git a/drivers/net/usb/lan78xx.c b/drivers/net/usb/lan78xx.c
index acd1a71f1698..e524777e611c 100644
--- a/drivers/net/usb/lan78xx.c
+++ b/drivers/net/usb/lan78xx.c
@@ -1287,8 +1287,12 @@ static void lan78xx_status(struct lan78xx_net *dev, struct urb *urb)
 		netif_dbg(dev, link, dev->net, "PHY INTR: 0x%08x\n", intdata);
 		lan78xx_defer_kevent(dev, EVENT_LINK_RESET);
 
-		if (dev->domain_data.phyirq > 0)
+		if (dev->domain_data.phyirq > 0) {
+			unsigned long flags;
+			local_irq_save(flags);
 			generic_handle_irq(dev->domain_data.phyirq);
+			local_irq_restore(flags);
+		}
 	} else
 		netdev_warn(dev->net,
 			    "unexpected interrupt: 0x%08x\n", intdata);
-- 
2.17.1

